### PR TITLE
V0.3.0

### PR DESCRIPTION
**Pull request by Khoroshevskyi**
New 0.3.0 version of pepdbagent
Main changes: 
- Big restructuring in pepdbagent functions.
- Restructured db: added new columns (`private`, `number_of_samples`, `submission_date`, `last_update_date`), and deleted annotation column (we didn't need it).
- Added new functionality: delete_project, update_item.

### CODE REVIEW

**Comment by nsheff** [comment]
- [ ] Add some comments to top of `models.py` to explain what it is
- [ ] Fix tutorial documentation
- [ ] Revisit docstrings. eg: https://github.com/pepkit/pepdbagent/blob/dev/pepdbagent/search.py#L17
- [ ] rename `_count_find_project`
- [ ] new tutorial for how the search works

File: pepdbagent/search.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+                        tag=res[2],
+                        number_of_samples=res[3],
+                        description=res[4],
+                        digest=res[5],
+                        is_private=res[6],
+                        last_update_date=str(res[8]),
+                        submission_date=str(res[7]),
+                    )
+                )
+        except KeyError:
```
Comment (nleroy917) [inline pepdbagent/search.py:171]: Out of curiosity, what would cause a `KeyError` here?

File: pepdbagent/search.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+                        number_of_projects=res[1],
+                        number_of_samples=res[2],
+                    )
+                )
+        except KeyError:
+            results_list = []
+
+        return results_list
+
+    def __run_sql_fetchall(self, sql_query: str, *argv) -> list:
```
Comment (nleroy917) [inline pepdbagent/search.py:241]: Might be over-engineering, but I notice that you use this same method in the `Connection` class. Might be worth making a `BaseConnection` class that has these sort of methods, and let `Connection` and `Search` inherit from it

**Review by nleroy917 — APPROVED** [review]
I think that this looks good. It already works well with pephub.

File: pepdbagent/search.py (hunk 2 above)
Comment (Khoroshevskyi) [reply pepdbagent/search.py:241]: I didn't know how to solve this problem, but now you gave me an idea, thanks!!
I try to write something like that

File: pepdbagent/search.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+            results=self._find_project(
+                namespace=namespace,
+                search_str=search_str,
+                admin=admin,
+                offset=offset,
+                limit=limit,
+            ),
+        )
+
+    def _get_number_of_found_projects(
```
Comment (nsheff) [inline pepdbagent/search.py:100]: Can you make this match the style of "_get_count_of_..." functions?

**Comment by nsheff** [comment]
It's not clear what the search functions return in the documentation: 
image

Shouldn't it return a dict?

File: pepdbagent/search.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+                search_str=search_str, admin_nsp=admin_list
+            ),
+            limit=limit,
+            offset=offset,
+            results=self._find_namespaces(
+                search_str=search_str, admin_nsp=admin_list, offset=offset, limit=limit
+            ),
+        )
+
+    def project(
```
Comment (nsheff) [inline pepdbagent/search.py:60]: It's not clear why there's: `con.get_project`, and `con.get_raw_project`, and `con.search.project`.

**Comment by nsheff** [comment]
I think you should rename the `Connection` object -- objects named `connection` or `con` are typically used to hold a connection to a database. In this case, though, it's not that -- it's more specific to a PEP database. This is confusing to me.

**Comment by nsheff** [comment]
It should be, maybe, `PEPDBConnection`, for example. Or, since it's not just a connection but does lots of stuff, ` `PEPDBAgent`.

Where is the `Agent` object in this, anyway?

File: pepdbagent/search.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+                        last_update_date=str(res[8]),
+                        submission_date=str(res[7]),
+                    )
+                )
+        except KeyError:
+            results_list = []
+
+        return results_list
+
+    def _get_count_of_found_namespaces(
```
Comment (nsheff) [inline pepdbagent/search.py:177]: why is the get count of found namespaces a different function from the get namespaces function?

**Review by nsheff — CHANGES_REQUESTED** [review]

File: pepdbagent/search.py (hunk 5 above)
Comment (nsheff) [reply pepdbagent/search.py:177]: I would suggest: 

Write a single function that runs 2 queries: 1. retrieve the objects of interest; 2. counts them. Put these into a standardized output structure, and return that.

**Comment by Khoroshevskyi** [comment]
> I don't understand why you made Connection and Search two different subclasses of BaseConnection.

I did it because we had duplicated function in two classes. So Base class holds this functions.

> 
> Wouldn't you just have an Agent class with a connection, that held instances of the sub-classes?

If I understood you correctly: You suggest to have all methods in one function, right? 
I divided this functionality for two classes, to make code more understandable, and to not mix all functions.

File: pepdbagent/search.py (hunk 5 above)
Comment (nleroy917) [reply pepdbagent/search.py:183]: > I don't understand why you made Connection and Search two different subclasses of BaseConnection.

This was my recommendation. I noticed he was duplicating code for the `Search` and `Connection` class, so I recommended make a base class that holds the `_run_sql_fetchall` type code.

**Comment by nleroy917** [comment]
> I don't understand why you made Connection and Search two different subclasses of BaseConnection.

This was my recommendation. I noticed he was duplicating code for the `Search` and `Connection` class, so I recommended make a base class that holds the `_run_sql_fetchall` type code.

**Comment by nleroy917** [comment]
@Khoroshevskyi @nsheff 
### Notes from January 12th, 2023 meeting:
- Change the name of the class:
  - Keep package name as `pepdbagent` and change class to `PEPDatabaseAgent`

**Comment by nsheff** [comment]
```
db_agent.namespace.get(limit=x, offset=0)  # all namespaces, { limit: x, offset: y, count: n, items: {...}}
db_agent.namespace.get(query="<query_string>", limit=x, offset=0)  # search namespaces, return { limit: x, offset: y, count: n, items: {...}}

db_agent.annotation.get(namespace=None, project_name=None, tag=None, limit=x, offset=0, query="<str>") # returns { limit: x, offset: y, count: n, items: {...}}
db_agent.annotation.get_by_rp(registry_paths, limit=x, offset=0) # { limit: x, offset: y, count: n, items: {...}}

db_agent.project.get_by_rp(registry_path) #
db_agent.project.get(namespace, project_name, tag)
```

**Pull request merged**